### ISSUE TITLE
[F#] Implement new Concept Exercise: numbers

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#807]
This issue describes how to implement the `strings` concept exercise for the F# track.

## Goal

The goal of this exercise is to teach the student how the Concept of Numbers is implemented in F#. It will introduce this concept through the two most common numeric types in F#: `int` (whole number) and `double` (floating-point number).

## Learning objectives

- Know of the existence of the two most commonly used number types, `int` and `double`.
- Understand that an `int` represents whole numbers, and a `double` represents floating-point numbers.
- Know of basic operators such as multiplication and comparison.
- Know how to convert from one numeric type to another; know that one always has to be explicit.
- Know how to conditionally execute code using an `if/elif/else` statement.

## Out of scope

- Any other numeric types besides `int` and `double` (so no `float`, `byte`, etc.).
- Parsing a `string` to an `int` or `double`.
- Converting an `int` or `double` to a `string`.
- Enumerating a string
- Overflows.

## Concepts

The Concepts this exercise unlocks are:

- `numbers-basic`: know of the existence of the two most commonly used number types, `int` and `double`; understand that the former represents whole numbers, and the latter floating-point numbers; now of basic operators such as multiplication and comparison.
- `type-conversion-numbers`: know how to convert from one numeric type to another using conversion operators.
- `conditionals-if`: know how to conditionally execute code using an `if` expressions.

## Prerequisites

There should be no prerequisites.

### Hints

- [Conditional-expressions][conditional-expressions]
- [Conversion operators][conversion-operators]

### After

- [Conditional-expressions][conditional-expressions]
- [Conversion operators][conversion-operators]

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement the concept exercise, the following files must be created:

<pre>
languages
└── fsharp
    └── exercises
        └── concept
            └── numbers
                ├── .docs
                |   ├── instructions.md
                |   ├── introduction.md
                |   ├── hints.md
                |   └── after.md
                ├── .meta
                |   ├── config.json
                |   |── design.md
                |   └── Example.fs
                ├── Numbers.fs
                ├── Numbers.fsproj
                └── NumbersTests.fs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should be explicit about what the student should learn from the exercise, and provide a short, concise introduction to the concept(s). The aim is to give the student just enough context to figure things out themselves and solve the exercise, as research has shown that self-discovery is the most effective learning experience. Mentioning technical terms that the student can Google if they so want, is preferable over including any code samples or an extensive description. For example we might describe a string as a "Sequence of Unicode characters" or a "series of bytes" or "an object". Unless the student needs to understand the details of what those mean to be able to solve the exercise we should not give more info in this introduction - instead allowing the student to Google, ignore, or map their existing knowledge.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

These files are also all described in the [concept exercises document][docs-concept-exercises].

## Step 5: update languages/fsharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "numbers",
        "uuid": "6ea2765e-5885-11ea-82b4-0242ac130003",
        "concepts": [
          "numbers-basic",
          "type-conversion-numbers",
          "conditionals-if"
        ],
        "prerequisites": []
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the F# track:

- `Numbers.fs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `Numbers.fsproj`: the F# project file.
- `NumbersTests.fs`: the test suite.
- `.meta/Example.fs`: an example implementation that passes all the tests. The project file should _not_ include this file.

## Step 7: update the general concept document

Add the exercise to the [concept's shared document's][reference] `## Implementations` section ([example](https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations)).

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][meta-design]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][meta-config.json]).

## Step 11: add reference documentation

This exercise could benefit from having a reference document written to introduce basic string concatenation in F#. See [this issue][string-concat-issue]

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[reference]: https://github.com/exercism/v3/blob/master/languages/fsharp/reference/README.md#reference-docs
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[meta-design]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/enums-advanced/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/enums-advanced/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[conditional-expression]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/conditional-expressions-if-then-else
[conversion-operators]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/casting-and-conversions#arithmetic-types
